### DISCUSSION TITLE
`__init_subclass__` and `__fields__`/`model_fields` in v2

### DISCUSSION THREAD

**Discussion by Kircheneer** [pydantic/pydantic#7177] [Question] [answered]
In the [diffsync](https://github.com/networktocode/diffsync) library, we use the `__init_subclass__` hook to verify whether the fields defined on the model are part of a private field defined on the model. Simple example:

```python
class Person(DiffSyncModel):
    _identifiers = ("name",)
    _attributes = ("age",)

    name: str
    age: int
```

We use the `__init_subclass__` hook in `DiffSyncModel` (which is a `pydantic.BaseModel` subclass) in order to verify that
- all fields referenced by `_identifiers` and `_attributes` exist
- no field is in both categories

With v2, this is no longer possible like this because neither the deprecated `Person.__fields__` nor the new `Person.model_fields` report the existence of `name` and `age` when `__init_subclass__` is executed. If I drop the validation and allow the subclass to be instantiated, I can see that the fields are present.

Any pointers on other way I can go about accessing this information? I have looked at `__annotations__`, which would probably work with some effort and `schema`, which would also probably work with some effort, but both don't fell 100% right to me.

**Comment by Kircheneer** [comment, marked as answer]
Solution found - use `__pydantic_init_subclass__`

**Reply by msto** [reply]
Thank you!
